File: OOPsand/OOPsand/ViewController.m
//
//  ViewController.m
//  OOPsand
//
//  Created by Chad Slater on 10/20/14.
//  Copyright (c) 2014 Chad Slater. All rights reserved.
//

#import "ViewController.h"
#import "NSObject+Video_Games.h"



@implementation ViewController



- (void)viewDidLoad {
    [super viewDidLoad];
    
   
}

- (IBAction)getGame{
    
    Video_Games *game = [[Video_Games alloc] init];
    game.developer = @"Bungie";
    game.year = 2001;
    game.title = @"Halo: Combat Evolved";
    
    
    Video_Games *game2 = [[Video_Games alloc] init];
    [game2 setTitle:@"Mass Effect"];
    [game2 setDeveloper:@"BioWare"];
    [game2 setYear:2007];
    
    NSArray *games = [[NSArray alloc] initWithObjects:game, game2, nil];
    
    Video_Games *halo = games[0];
    Video_Games *massEffect = games[1];
    

    switch (self.getGameSegmented.selectedSegmentIndex)
    {
            
            
        case 0:
            self.gameYearLabel.text = [NSString stringWithFormat:@"Year Released: %d", halo.year];
            self.gameDeveloperLabel.text = [NSString stringWithFormat:@"Developed By: %@",halo.developer];
            break;
        case 1: self.gameDeveloperLabel.text = [NSString stringWithFormat:@"Developed By: %@", massEffect.developer];
             self.gameYearLabel.text = [NSString stringWithFormat:@"Year Released: %d", massEffect.year];
            break;
        default:
            break; 
    }
    
    
    
    
    
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}


@end

File: DesignPatterning/DecoratorPattern/DecoratorPattern/ViewController.swift
//
//  ViewController.swift
//  DecoratorPattern
//
//  Created by Chad Slater on 11/24/14.
//  Copyright (c) 2014 Chad Slater. All rights reserved.
//

import UIKit

protocol MacBook {
    func getCost() -> Double
    func getScreenSize() -> String
    func getMemoryAmount() -> String
}

class ViewController: UIViewController {
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        
        class RegularMacBook: MacBook {
            func getCost() -> Double {
                return 1000.0
            }
            
            func getScreenSize() -> String {
                return "13 Inch"
           
            }
            func getMemoryAmount() -> String {
                return "8GB"
            }
        }
    
        class MacBookDecorator: MacBook {
            private let decoratedMacBook: MacBook
            
            required init(decoratedMacBook: MacBook) {
                self.decoratedMacBook = decoratedMacBook
            }
            
            func getCost() -> Double {
                return decoratedMacBook.getCost()
            }
            
            func getScreenSize() -> String {
                return decoratedMacBook.getScreenSize()
            }
            func getMemoryAmount() -> String {
                return decoratedMacBook.getMemoryAmount()
            }
        }
        
        class Pro: MacBookDecorator {
            required init(decoratedMacBook: MacBook) {
                super.init(decoratedMacBook: decoratedMacBook)
            }
            
            override func getCost() -> Double {
                return super.getCost() + 500.0
            }
            
            override func getScreenSize() -> String {
                return super.getScreenSize() + ", " + " + 2 inches"
            }
            override func getMemoryAmount() -> String {
                return super.getMemoryAmount() + ", " + " + 8GB"
            }
        }
        
        class Air: MacBookDecorator {
            required init(decoratedMacBook: MacBook) {
                super.init(decoratedMacBook: decoratedMacBook)
            }
            
            override func getCost() -> Double {
                return super.getCost() - 400.0
            }
            
            override func getScreenSize() -> String {
                return super.getScreenSize() + ", " + " - 2 inches"
            }
            override func getMemoryAmount() -> String {
                return super.getMemoryAmount() + ", " + " - 4GB"
            }
        }
        
        
        var aMacBook: MacBook = RegularMacBook()
        println("Cost : \(aMacBook.getCost()); Screen Size: \(aMacBook.getScreenSize()) Memory Amount: \(aMacBook.getMemoryAmount())")
        aMacBook = Pro(decoratedMacBook: aMacBook)
        println("Cost : \(aMacBook.getCost()); Screen Size: \(aMacBook.getScreenSize()) Memory Amount: \(aMacBook.getMemoryAmount())")
        aMacBook = Air(decoratedMacBook: aMacBook)
        println("Cost : \(aMacBook.getCost()); Screen Size: \(aMacBook.getScreenSize()) Memory Amount: \(aMacBook.getMemoryAmount())")
        
    }
    
    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
    
    
}

